GCTC-NTGC/gc-digital-talent
# Hydrogen 2.0.0-beta.38

## 👋 Introduction

This PR sets the stage for Hydrogen's official `2.0.0` release by ensuring any possible bugs are squashed.

Latest Chromatic build: https://www.chromatic.com/pullrequest?appId=61e099a1bb1465003a73d3ce&number=5309&view=changes

## 🕵️ Details

This PR upgrades Hydrogen from `beta.33` to `beta.38` and *does include breaking changes* which have been summarized below. Bugs found as a part of this PR will inform any remaining Hydrogen beta releases before `2.0.0` is published.

### New features

- Previously, the use of configured values was limited to Hydrogen supported properties. With the addition of CSS property support in `beta.30`, it was always the intent to allow the use of configured values inside of ANY property that made sense. This means that you can now use values such as `primary` inside of any CSS property that accepts colors (e.g. `data-h2-text-shadow="base(0px 3px 1px primary.dark)"`)
  - This extends to all configured values, include space units (e.g. `x2`), meaning you can now use the vertical rhythm values in properties that would normally accept pixels or `rem` units.
  - If you find a property that doesn't support a configured value that you would expect, please let me know and I'll review adding it before the `2.0.0` release.
- The snippets file has been updated to reflect many more available properties, but is not exhaustive to avoid clutter in the snippet peek function. If something is missing that you would find useful, just shout.
- Dark mode's automatic functionality has been enhanced so that any color modifier that Hydrogen automatically generates swaps with its tint/shade counterpart without any styling intervention. This means that when you use `primary.dark` in light mode, it will automatically flip to `primary.light` for you in dark mode without any effort.

### Breaking changes

- Color modifiers (e.g. `lighter` or `dark`) are now only processed as suffixes to the color name
  - For example, `primary.dark`; `dark.primary` will no longer work. This is the result of Hydrogen enabling the use of configured colors in ALL CSS properties that accept color options and has to do with the expression used to find configured keys.
- A handful of properties now follow their CSS counterparts to bring Hydrogen in line with CSS syntax
  - `border` used to accept 4 comma separated values, but now accepts 3 values separated by spaces (e.g. `border="base(left, 1px, solid, red)"` now becomes `border-left="base(1px solid red)"`
  - `background-color` has lost support for configured gradient values; `background` should be used instead
  - `gap`'s options have been reversed to match its native CSS syntax
  - `offset` is now `location` to enable support for vanilla CSS offset
  - `visibility` has been swapped to its vanilla CSS visibility, but a new custom property `visually-hidden` has been added to retain the original's functionality

## 🧪 Testing

Assist reviewers with steps they can take to test that the PR does what it says it does.

1. Reset your dependencies by running the setup script in `maintenance` to ensure you have the latest version of Hydrogen installed
2. Run the development and watch scripts to ensure that no Hydrogen errors are returned
3. Run storybook to ensure that everything is displaying as expected
4. Try using some of the new features!
